src/java/boa/compiler/visitors/analysis/LocalMayAliasAnalysis.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright 2014, Hridesh Rajan, Robert Dyer, 
+ *                 and Iowa State University of Science and Technology
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package boa.compiler.visitors.analysis;
+
+import java.util.*;
+
+import boa.compiler.ast.Call;
+import boa.compiler.ast.Comparison;
+import boa.compiler.ast.Component;
+import boa.compiler.ast.Composite;
+import boa.compiler.ast.Conjunction;
+import boa.compiler.ast.Factor;
+import boa.compiler.ast.Identifier;
+import boa.compiler.ast.Index;
+import boa.compiler.ast.Node;
+import boa.compiler.ast.Operand;
+import boa.compiler.ast.Pair;
+import boa.compiler.ast.Selector;
+import boa.compiler.ast.Term;
+import boa.compiler.ast.UnaryFactor;
+import boa.compiler.ast.expressions.Expression;
+import boa.compiler.ast.statements.AssignmentStatement;
+import boa.compiler.ast.statements.VarDeclStatement;
+import boa.types.*;
+import boa.compiler.visitors.*;
+
+/**
+ * @author rramu
+ */
+public class LocalMayAliasAnalysis extends AbstractVisitorNoArg {
+	HashSet<Identifier> aliastSet = new HashSet<Identifier>();
+	Identifier lastDefVariable;
+	boolean cloneFound = false;
+	protected final IdentifierFindingVisitor idFinder = new IdentifierFindingVisitor();
+	protected final CallFindingVisitor callFinder = new CallFindingVisitor();
+
+	protected class CallFindingVisitor extends AbstractVisitorNoArg {
+		protected boolean isCall;
+
+		public boolean isCall() {
+			return isCall;
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void initialize() {
+			super.initialize();
+			isCall = false;
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void visit(final Factor n) {
+			for (final Node node : n.getOps()) {
+				isCall = false;
+				node.accept(this);
+			}
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void visit(final Call n) {
+			isCall = true;
+		}
+	}
+
+	protected class IdentifierFindingVisitor extends AbstractVisitorNoArg {
+		protected final Set<String> names = new HashSet<String>();
+
+		public Set<String> getNames() {
+			return names;
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		protected void initialize() {
+			names.clear();
+		}
+
+		/** {@inheritDoc} */
+		@Override
+		public void visit(final Identifier n) {
+			names.add(n.getToken());
+		}
+	}
+
+	public final void dfs(final Node node, java.util.HashMap<Integer,String> nodeVisitStatus) {
```

Comment:
I'm a bit confused of the point of this (and the need for the nodeVisitStatus map).  The default visitor **does** a DFS visit of the tree.